### PR TITLE
Add alpha command to kubectl

### PR DESCRIPTION
**Pull request by verb**
Also allow new commands to disable themselves by returning a nil value. This can be used to disable commands based on feature gates.

**What this PR does / why we need it**: Method of enabling alpha functionality in kubectl

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: ref #45922

**Special notes for your reviewer**: Part of a discussion in #45922 with @pwittrock

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by verb** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

File: pkg/kubectl/cmd/cmd.go
Comment (xiangpengzhao) [inline pkg/kubectl/cmd/cmd.go]: Document how to set this env?

File: pkg/kubectl/cmd/cmd.go
Comment (xiangpengzhao) [inline pkg/kubectl/cmd/cmd.go]: provide some way to get KnownFeatures and whether they are enabled or not?

File: pkg/kubectl/cmd/cmd.go
Comment (xiangpengzhao) [inline pkg/kubectl/cmd/cmd.go]: or give some help info via `kubectl --help`

File: pkg/kubectl/cmd/cmd.go
Comment (xiangpengzhao) [inline pkg/kubectl/cmd/cmd.go]: What if users run a disabled command?

File: pkg/kubectl/cmd/cmd.go
Comment (xiangpengzhao) [inline pkg/kubectl/cmd/cmd.go]: how about set feature gates via a subcommand or flag instead of setting env vars?

Comment (verb) [inline pkg/kubectl/cmd/cmd.go]: Thanks for asking this, it made me realize that I hadn't actually Submitted my comment explaining my reasoning and it was just hanging out in a text box in another tab waiting for me to close my browser. Here it is: https://github.com/kubernetes/kubernetes/issues/45922#issuecomment-302895410

There's a bit of a catch-22 with setting feature gates on kubectl itself. If we set them with a flag then the gates won't be available prior to flag parsing and so can't gate flags or the command tree.

We definitely should document this, but first let's reach consensus on how the feature gates will be implemented.

File: pkg/kubectl/cmd/cmd.go
Comment (verb) [inline pkg/kubectl/cmd/cmd.go]: Good idea. I'll include this in `kubectl --help` as you suggested.

File: pkg/kubectl/cmd/cmd.go
Comment (verb) [inline pkg/kubectl/cmd/cmd.go]: When a command disables itself by returning `nil` it's as if the command doesn't exist, which is intentional. My goal is that alpha features don't influence the behavior of `kubectl` in any way unless they're enabled.

**Comment by sttts** [comment]
We have had the discussion a number of times whether clients should depend on apiserver features. IMO if we want feature gates in kubectl at all, we should make them independent from the apiserver ones.

/cc @smarterclayton

File: pkg/kubectl/cmd/cmd.go
Comment (xiangpengzhao) [inline pkg/kubectl/cmd/cmd.go]: @verb Thanks for the explanation!

**Comment by verb** [comment]
@sttts Sure, it would be easy to use the feature gate framework with an independent configuration. I rather like the feature gate framework, but I'm not tied to it. If we do use feature gates do you like the environment-based approach?

I think the alternative would have to be a `kubectl alpha` subcommand would could segregate the alpha commands but not hide them entirely.

File: pkg/kubectl/cmd/cmd.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/cmd.go]: If this is an error it should be named `err`, no?

**Review by alexandercampbell — COMMENTED** [review]
Code looks fine to me. I'll defer to the existing discussion on whether this is the correct path to take.

File: pkg/kubectl/cmd/cmd.go
Comment (verb) [inline pkg/kubectl/cmd/cmd.go]: Normally it would, but I want to print to the `io.Writer` with the same name from the enclosing scope.

Comment (alexandercampbell) [inline pkg/kubectl/cmd/cmd.go]: Gotcha, thanks!

**Comment by verb** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by verb** [comment]
@sttts @alexandercampbell

I tried a different approach and created a `kubectl alpha` subcommand based on your feedback. I think I may like this approach better. The command is hidden if there are no alpha commands available. The command for my alpha feature would be come `kubectl alpha debug`.

Right now alpha commands are visible if they exist, but it would be easy to hide them further.

**Comment by xiangpengzhao** [comment]
@verb pull-kubernetes-verify failed due to:

`I0531 19:07:33.642] FAILED   hack/make-rules/../../hack/verify-generated-docs.sh	38s`

**Comment by sttts** [comment]
I like the alpha command more than the env var 👍

File: pkg/kubectl/cmd/alpha.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/alpha.go]: Where is `alphaGroups` defined?

File: pkg/kubectl/cmd/cmd.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/cmd.go]: Comment might be nice here. Something like

```go
// If there are no alpha commands in this build, hide the "alpha" subcommand.
```

Comment (verb) [inline pkg/kubectl/cmd/cmd.go]: Good call, done.

**Comment by fejta** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
ref https://github.com/kubernetes/test-infra/pull/2931

File: pkg/kubectl/cmd/alpha.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/alpha.go]: Maybe this should be:

```go
Use: "alpha SUBCOMMAND",
```

To be consistent with other commands in the repo:

```
cmd/certificates.go:35:		Use:   "certificate SUBCOMMAND",
cmd/set/set.go:37:		Use:   "set SUBCOMMAND",
cmd/rollout/rollout.go:51:		Use:     "rollout SUBCOMMAND",
cmd/config/config.go:39:		Use:   "config SUBCOMMAND",

```

File: pkg/kubectl/cmd/alpha.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/alpha.go]: You don't need the underscores here.

**Review by alexandercampbell — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/alpha.go
Comment (verb) [inline pkg/kubectl/cmd/alpha.go]: I'd be happy to do that, but alpha is a little bit different from set/rollout/config as its more of a container for commands rather than a command with subcommands. For example, `config get` and `config set` are related subcommands that operate on `config`, but `alpha debug` and `alpha foo` are unrelated other than both being in an alpha state.

Also, the commands in alpha could themselves have subcommands, which would make the "Usage: kubectl alpha SUBCOMMAND [options]" not quite correct. 

As its coded now, without a run function, Usage will be omitted from help and look like this:

```
These commands correspond to alpha features that are not enabled in Kubernetes clusters by default.                                                                            
                                                                                                                                                                               
Available Commands:                                                                                                                                                            
  debug       Run a debug container in a pod                                                                                                                                   
                                                                                                                                                                               
Use "kubectl <command> --help" for more information about a given command.   
```

If we were to give it a default run function and set Use to  "alpha COMMAND", then the help would be:
```
These commands correspond to alpha features that are not enabled in Kubernetes clusters by default.                                                                            
                                                                                                                                                                               
Available Commands:                                                                                                                                                            
  debug       Run a debug container in a pod                                                                                                                                   
                                                                                                                                                                               
Usage:                                                                                                                                                                         
  kubectl alpha COMMAND [options]                                                                                                                                           
                                                                                                                                                                               
Use "kubectl <command> --help" for more information about a given command.                                                                                                     
Use "kubectl options" for a list of global command-line options (applies to all commands). 
```

**Comment by pwittrock** [comment]
@verb would you squash before I approve?

**Comment by pwittrock** [comment]
Looks like the merge bot was over eager about the approve label

**Comment by verb** [comment]
squashed!

**Pull request merged**
